src/scmdata/timeseries.py

```diff
@@ -0,0 +1,209 @@
+"""
+TimeSeries handling
+
+Functionality for handling and storing individual time-series
+"""
+
+import copy
+import datetime as dt
+import functools
+from typing import Any, Callable, List, Union
+
+import numpy as np
+import xarray as xr
+from xarray.core.ops import inject_binary_ops
+
+from scmdata.time import TimeseriesConverter
+
+
+class _Counter:
+    def __init__(self):
+        self.count = 0
+
+    def __call__(self):
+        val = self.count
+        self.count += 1
+
+        return val
+
+    def reset(self):
+        self.count = 0
+
+
+default_name = _Counter()
+
+
+class TimeSeries:
+    """
+    A 1D time-series with metadata
+
+    Proxies a xarray.DataArray with a single time dimension
+    """
+
+    def __init__(self, data, **kwargs):
+        values = np.asarray(data)
+
+        if values.ndim != 1:
+            raise ValueError("TimeSeries must be 1d")
+        if isinstance(data, xr.DataArray):
+            self._data = data
+        else:
+            # Auto incrementing name
+            if "name" not in kwargs:
+                kwargs["name"] = default_name()
+
+            self._data = xr.DataArray(values, **kwargs)
+
+    def __repr__(self):
+        return self._data.__repr__()
+
+    @property
+    def name(self):
+        """
+        Timeseries name
+
+        If no name was provided this will be an automatically incrementing number
+        """
+        return self._data.name
+
+    def __len__(self):
+        """
+        Length of the time-series (number of time points)
+
+        Returns
+        -------
+        int
+        """
+        return len(self._data["time"])
+
+    def copy(self):
+        """
+        Create a deep copy of the timeseries.
+
+        Any further modifications to the :obj:`Timeseries` returned copy will not be reflecting
+        in the current :obj:`Timeseries`
+
+        Returns
+        -------
+        :obj:`Timeseries`
+        """
+        return copy.deepcopy(self)
+
+    @property
+    def meta(self):
```

Comment:
It is in the same format as pyam's meta table which is why we had it in ScmDataFrame

Reply:
Just saw this is referring to the timeseries. I'd be happy to change to `metadata` which is aggregated into the `meta` property in `ScmRun`

Reply:
> It is in the same format as pyam's meta table which is why we had it in ScmDataFrame

I spoke with the pyam developers a bit more a while back and it turns out this isn't actually right (more below). So I think the first question is, "Do we want to always have a comparison to pyam or do we just not worry about their data format anymore beyond being able to initialise from an `IamDataFrame` and convert to an `IamDataFrame`?" (For what it's worth, I think we don't worry about the comparison as we'll spend too much effort.)

The pyam `meta` table isn't what we thought it was. Full details are [here](https://pyam-iamc.readthedocs.io/en/stable/data.html) and the PR which resulted from my discussions with the pyam devs is [here](https://github.com/IAMconsortium/pyam/pull/286). The shortish summary is that the pyam `meta` table is indexed by model-scenario combinations  i.e. the information in the `meta` table for all timeseries related to a given model-scenario must be identical. This means, for example, that units don't go in the `meta` table, rather they're part of the `data` table. In contrast, our metadata is timeseries specific so e.g. units do go in the `meta` table. It's that difference which made me think we might want to use a different name. 

The reason I suggested `metadata` is because it's a bit more explicit, we have our data and then we have metadata about the data (on a timeseries by timeseries level). `meta` as a shorthand is also good, we just need to be clear in our own minds that our `meta` is a completely different concept to what pyam has.

Reply:
As discussed, we'll keep it as meta, but with some documentation mentioning the difference to pyam

Reply:
Addressed by #41